client/my-sites/email/email-providers-stacked-comparison/email-provider-stacked-card/email-provider-generic-form.tsx

```diff
@@ -0,0 +1,868 @@
+			: mailBoxError,
+} );
+
+/**
+ * Retrieves the specified user after applying a callback to all of its fields.
+ *
+ * @param {object} user - user with a list of fields
+ * @param {Array} optionalFields - list of optional fields (not required fields)
+ * @param {Function} callback - function to call for each field
+ */
+const mapFieldValues = (
+	user: GenericNewUser,
+	optionalFields: string[],
+	callback: any
+): GenericNewUser =>
+	mapValues( user, ( fieldValue, fieldName ) => {
+		if ( 'uuid' === fieldName || optionalFields.some( ( field: string ) => field === fieldName ) ) {
+			return fieldValue;
+		}
+
+		return callback( fieldValue, fieldName, user );
+	} );
+
+/*
+ * Adds a new error to the specified email field if it has no errors and contains invalid characters.
+ */
+const validEmailCharacterField = ( {
+	value,
+	error,
+}: GenericNewUserField ): GenericNewUserField => ( {
+	value,
+	error:
+		! error && ! /^[0-9a-z_'-](\.?[0-9a-z_'-])*$/i.test( value )
+			? translate(
+					'Only numbers, letters, dashes, underscores, apostrophes and periods are allowed.'
+			  )
+			: error,
+} );
+
+/*
+ * Adds a new error to the specified field if it has no errors and is more than sixty characters.
+ */
+const sixtyCharacterField = ( { value, error }: GenericNewUserField ): GenericNewUserField => ( {
+	value,
+	error:
+		! error && 60 < value.length
+			? translate( "This field can't be longer than %s characters.", {
+					args: '60',
+			  } )
+			: error,
+} );
+
+/*
+ * Adds a new error if the specified mailbox field has no errors and appears more than once in the provided map.
+ */
+const validateNewUserMailboxIsUnique = (
+	{ value: mailBox, error: previousError }: GenericNewUserField,
+	mailboxesByCount: { [ mailbox: string ]: number }
+) => ( {
+	value: mailBox,
+	error:
+		mailboxesByCount[ mailBox ] > 1
+			? translate( 'Please use a unique mailbox for each user.' )
+			: previousError,
+} );
+
+/*
+ * Adds a duplicate error to each mailBox with a duplicate mailbox
+ */
+const validateNewUsersAreUnique = ( users: GenericNewUser[] ): GenericNewUser[] => {
+	const mailboxesByCount: { [ mailbox: string ]: number } = countBy(
+		users.map( ( { mailBox: { value: mailBox } } ) => mailBox )
+	);
+
+	return users.map(
+		( {
+			uuid,
+			alternativeEmail,
+			fullName,
+			email,
+			domain,
+			isAdmin,
+			mailBox,
+			firstName,
+			lastName,
+			password,
+		} ) => ( {
+			uuid,
+			alternativeEmail,
+			email,
+			firstName,
+			fullName,
+			lastName,
+			domain,
+			isAdmin,
+			mailBox: validateNewUserMailboxIsUnique( mailBox, mailboxesByCount ),
+			password,
+		} )
+	);
+};
+
+/*
+ * Validates the specified password field.
+ *
+ * @see https://support.google.com/accounts/answer/32040 for requirements
+ */
+export const validatePasswordField = (
+	{ value, error }: GenericNewUserField,
+	minimumLength = 12
+): GenericNewUserField => {
+	if ( ! error && minimumLength > value.length ) {
```

Comment:
Instead of checking `! error` in every condition, might it be simpler to return early when `error` is truthy?
```suggestion
	if ( error ) {
		return { value, error };
	}

	if ( minimumLength > value.length ) {
		...
```